keps/sig-storage/20200217-namespace-transfer.md

```diff
@@ -0,0 +1,283 @@
+- The API should be compatible with as many existing storage provisioners as possible.
+
+### Non-Goals
+
+- Transfer of resources other than PersistentVolumeClaims will not be discussed.
+- Implementation details only discussed in the context of influencing the API definition.
+
+## Proposal
+
+`StorageTransferRequest` resource:
+
+```yaml
+apiVersion: v1alpha1
+kind: StorageTransferRequest
+metadata:
+  name: transfer-foo
+  namespace: target-namespace
+spec:
+  source:
+    name: foo
+    namespace: source-namespace
+    kind: PersistentVolumeClaim
+  targetName: bar #this is optional
+```
+
+`StorageTransferApproval` resource:
+
+```yaml
+apiVersion: v1alpha1
+kind: StorageTransferApproval
+metadata:
+  name: approve-foo
+  namespace: source-namespace
+spec:
+  source:
+    name: foo
+    kind: PersistentVolumeClaim
+  targetNamespace: target-namespace
+```
+
+When matching `StorageTransferRequest` and `StorageTransferApproval` resources are detected, the transfer process is initiated.
+
+Matching requests/approvals is similar to the process for binding PersistentVolumes with PersistentVolumeClaims.
+If they are not explicitly bound by the user, a controller will attempt to pair them by looking for matching resource types/names.
+
+Once matching is complete, a controller will begin the transfer process which does the following (not necessarily in order):
+
+- Make sure the associated PersistentVolume is not deleted/recycled during the transfer by setting reclaim policy to `Retain` (if not already).
+- Delete the PVC `source-namespace\foo`.
+- Bind The PersistentVolume to `target-namespace\bar`.
+- Create the PVC `target-namespace\bar` by copying the spec of `source-namespace\foo` and setting `spec.volumeName` appropriately.
+
+The last step makes it possible for the API to be compatible with the most provisioners
+as they will ignore a PVC that is already bound.  In [previous](https://github.com/kubernetes/enhancements/pull/1112) namespace transfer KEPs, the target PVC is created by the user.  Although, this will probably work with the CSI external provisioner, it will not work with others (like static local volume).
+
+Users can monitor the status of a transfer by querying for the existence of the bound target PVC or checking the `status.complete` property of `StorageTransferRequest`.
+
+`StorageTransferRequest` object:
+
+```golang
+type StorageTransferRequest struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  Spec StorageTransferRequestSpec
+  Status *StorageTransferRequestStatus
+}
+```
+
+```golang
+type StorageTransferRequestSpec struct {
+  Source *corev1.ObjectReference
+
+  // ApprovalName is set by the controller when requests/approvals are matched
+  // or by the user when manual binding is preferred
+  ApprovalName string
+
+  // TargetName allows for the source and target resources to have different names
+  // It is optional
+  TargetName *string
```

Comment:
It may be better to always keep this field populated (which would simplify checking in the code)

Reply:
Hi @tedyu - thanks for taking a look.  Yes, your suggestion would definitely make the implementation simpler.  But I wonder about the case in which the user doesn't have access to the source namespace and some coordination with an admin is necessary.  Now one more piece of information is required to be communicated.  Maybe it's not a big deal, but at this point I kind of feel like the benefit outweighs the cost.

Reply:
Since the transfer is secure, I wonder under what scenario the user doesn't have access to the source namespace.

Reply:
A hypothetical case could be something like:  Department A needs a copy of a database (PVC) owned by department B.  Each department operates only in their own namespace.  The admins for A and B can work together to facilitate the transfer rather than getting a cluster admin involved.